src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BedSearchService.kt

```diff
@@ -118,10 +123,33 @@ class BedSearchService(
 
         val results = candidateResults.filter { !bedsWithABookingInTurnaround.contains(it.bedId) }
 
+        val distinctIds = candidateResults.map { it.premisesId }.distinct()
+        val overlappedBookings = bookingRepository.findAllByPremisesIdsAndOverLappingDate(distinctIds, startDate, endDate)
+        val groupedOverlappedBookings = overlappedBookings
+          .map { transformBookingToOverlapResult(it, startDate, endDate) }
+          .groupBy { it.premisesId }
+
+        results.forEach {
+          val overlappingBookings = groupedOverlappedBookings[it.premisesId]?.toList() ?: listOf()
+          it.overlaps.addAll(overlappingBookings)
+        }
+
         return@validated success(
           results,
         )
       },
     )
   }
+
+  fun transformBookingToOverlapResult(booking: BookingEntity, startDate: LocalDate, endDate: LocalDate): TemporaryAccommodationBedSearchOverlapResult {
```

Comment:
I like the way you've pulled out some responsibility here 👍 

From what I've seen we tend to have long procedural methods with many responsibilities. As we move forwards in general (ie not now) I think it'd be great to extract more than transformations and get methods closer to the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle). The value I see is that a code base that's easier to read and reason about is easier to change.

Reply:
@danlivings-dxw gets at least half the credit on this one, but thanks :)
A big difference between the way dxw works and the way previous firms I've been at have worked is that whilst the latter sacrifices atomic commits for strict adherence to SOLID principles, dxw appears to do the opposite. (Personally, I lean further towards readability of code than readability of commits, as the code is the most-read aspect of any system by its developers and therefore ought to be self-documenting. I'd like to discuss this balance further in the future.)